### PR TITLE
Change ResponseMessage result to be `Either ResponseError a`

### LINKED ISSUE #191: Change response type to contain `Either a ResponseError`
> This seems to be another hint at the fact that we should change this to something like `Either a ResponseError` but that would be a breaking change so I would at least do that in a separate PR whereas this is only breaking if you rely on the behavior on invalid `json-rpc` messages.

_Originally posted by @cocreature in https://github.com/alanz/haskell-lsp/pull/189#issuecomment-541631689_

### PR DESCRIPTION
**Pull request by greenhat**
Close #191 

Todo:
- [x] add JSON roundtrip tests for `ResponseMessage` with valid values for `result` field (`string | number | boolean | object | null`);
- [x] check that invalid JSON (`result = null`(without `error` field) and both of them being absent) is handled properly;
- [x] remove `either` dependency (see https://github.com/alanz/haskell-lsp/pull/234#discussion_r395584136);
- [x] remove prefix in `ResponseMessage` fields (see https://github.com/alanz/haskell-lsp/pull/234#discussion_r395561185);

### CODE REVIEW

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs [hunk 1]
```diff
@@ -380,30 +381,58 @@ deriveJSON lspOptions{ fieldLabelModifier = customModifier } ''ResponseError
...
 data ResponseMessage a =
   ResponseMessage
-    { _jsonrpc :: Text
-    , _id      :: LspIdRsp
-    , _result  :: Maybe a
-    , _error   :: Maybe ResponseError
+    { _rmjsonrpc :: Text
+    , _rmid      :: LspIdRsp
+    , _rmresult  :: Either ResponseError a
     } deriving (Read,Show,Eq)
```
Comment (alanz) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: We have been naming the fields to correspond with the spec, with an underscore to allow idiomatic lens generation.  The `OverloadedRecordFields` extension allows us to do this.

So I think it would be better to stick with the current usage, i.e. get rid of the `rm` bit in the field names.

Comment (greenhat) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Thanks! Got it. I'll remove them. 
I've added `rm` prefixes because compiler errored on `toJson` :
```haskell
instance ToJSON a => ToJSON (ResponseMessage a) where
  toJSON r = object
    [ "jsonrpc" .= _rmjsonrpc r
    , "id" .= _rmid r
    , "result" .= rightToMaybe (_rmresult r)
    , "error" .= leftToMaybe (_rmresult r)
    ]
``` 
with functions (record fields) being ambigous. I'll find another way.

**Comment by greenhat** [comment]
@alanz Thanks for the review! It's work in progress though(draft PR). I'm not happy with roundtrip serialization tests and I need to clean up my code. :)

File: haskell-lsp-types/haskell-lsp-types.cabal [hunk 2]
```diff
@@ -58,6 +58,7 @@ library
                      , scientific
                      , text
                      , unordered-containers
+                     , either >= 5.0.1.1
```
Comment (bubba) [inline haskell-lsp-types/haskell-lsp-types.cabal]: Nitpicking here, if we’re only using those two helper functions from either would it be worthwhile to just write them ourselves as a `where` clause? Just to avoid adding more dependencies

Comment (greenhat) [reply haskell-lsp-types/haskell-lsp-types.cabal]: Thanks! Will do.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs (hunk 1 above)
Comment (greenhat) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Done.

File: haskell-lsp-types/haskell-lsp-types.cabal (hunk 2 above)
Comment (greenhat) [reply haskell-lsp-types/haskell-lsp-types.cabal]: Done.

**Comment by greenhat** [comment]
I finished with all I had planned. Please review and let me know if I missed anything.
